[WIP] Non streamable implementation of RFC 1951

clecat opened the pull request:
This work is currently in progress. It aims to provide a non streamable implementation of the RFC 1951.
The main use should be an increase with the performances as we do not need to save the general state of the compression/decompression every time.
The drawback is that the user needs to provide the entire object to compress/decompress in one single time.

There might still be some artifacts of the streamable implementation but they should be cleaned and removed as the time goes by.

To do:
- [x] Inflation implemented
- [x] Inflation largely tested
- [x] Deflation implemented
- [x] Deflation largely tested
- [x] zlib wrapper adapted
- ~gzip wrapper adapted~
- [x] Rewrite a clean commit history

dinosaure commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+      fill_bits d 8 ;
+      let nlen = pop_bits d 8 in
+      fill_bits d 8 ;
+      let nlen = pop_bits d 8 lsl 8 lor nlen in
```
I think the way to get `len` and `nlen` is bit overweight. I think, due to the last goal - re-align `hold` and `bits` - we can find a better way that this.

clecat replied:
Indeed, for now we can at least group them in a single call to 16 bits instead of 2 of 8

dinosaure replied:
According to this comment: https://github.com/mirage/decompress/pull/102/files#r543520015
All of these should be deleted and replaced by 2 simple `unsafe_get_uint16`.

dinosaure commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+      else
+        ( d.hold <- 0 ;
+          d.bits <- 0 ;
+          let m_len = min (min (i_rem d) len) (bigstring_length d.o - d.o_pos) in
```
Instead to try as far as we can to copy input to output, we should just return an error if `len > i_rem d || len > o_rem d`.

dinosaure commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+      in
+      let exception End in
+      let exception Invalid_distance in
+      let exception Invalid_distance_code in
```
`let exception`, as far I can say, is not yet optimized as a real C _jump_, we should define them outside the function - and say that these exceptions should be used only by `inflate`.

dinosaure commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+      ; mutable o_pos : int
+      ; w : WInf.t
+      ; mutable l : int (* literal / length *)
+      ; mutable d : int (* distance *)
```
`l` and `d` are useless here.

dinosaure commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+      inflate fixed_lit fixed_dist d
+
+    (* XXX(clecat): The table functions are almost a copy of the stream implementation, by
+       adapting their code, they should be easily merged *)
```
Please do.

clecat replied:
I'll push the other modifications for now and change that in a separate commit, so we keep clear tracks of it

dinosaure replied:
Seems like good enough, this part is not the most important to review.

dinosaure commented on lib/de.mli:
```diff
@@ -118,6 +118,32 @@ module Inf : sig
+
+    val pp_error: Format.formatter -> error -> unit
+
+    val inflate : src:bigstring -> dst:bigstring -> w:window -> (int * int, error * (int * int)) result
```
I think, in the case of `error`, we should just return an error - in that case, `src` is not correct and `dst` is compromise.

dinosaure reviewed (commented):
A first pass on the code, we should focus now on tests about the corpus (which gives to us some special cases).

clecat commented on lib/de.ml:
```diff
@@ -1382,6 +1382,306 @@ module Inf = struct
+
+    let flat d =
+      d.hold <- d.hold lsr (d.bits land 7) ;
+      d.bits <- d.bits land (lnot 7) ;
```
Those lines align the hold on 8 bits.
It can be found in the streamable implementation [there](https://github.com/mirage/decompress/blob/e21bd8dbfc10dd35abd9947902c613f7ec870dac/lib/de.ml#L832)

dinosaure replied:
Ok, so the situation is a bit more clear for me. According to [RFC1951](https://tools.ietf.org/html/rfc1951#section-3.2.4), `len` and `nlen` are aligned. At least, the worst situation for you is when you have more than 8 bits in your `hold`. In that case, a part of `len` is into `hold`.

I think and in your situation, the best is to re-update `d.i` (move back) if you have strictly more than 7 bits in your `hold` and then, do two simple `unsafe_get_uint16` to get `len` and `nlen`.

clecat commented on lib/de.ml:
```diff
@@ -1382,6 +1382,302 @@ module Inf = struct
+          then
+            ( let v = unsafe_get_uint8 d.w.WInf.raw off in
+              WInf.fill d.w v d.o d.o_pos len )
+          else
```
I believe this part of the code might also need a rework, as I didnt change it at all.
The output flushing related code should still be here

dinosaure replied:
A fast look tell me that we don't need to recalculate `len` according to `(d.o_len - d.o_pos)` but we should check that `l` is lower or equal to `(d.o_len - d.o_pos)` - otherwise, you raise an error.

Then, in any case at the end of this part of the code should call `length ()` directly without any `if`.

The use of `blit`/`pre`/`rst` is ok - and I think needed.

dinosaure commented on lib/de.ml:
```diff
@@ -1503,35 +1503,44 @@ module Inf = struct
 
     let inflate lit dist d =
+      let rec fill_bits d lit n =
+        if d.bits < n && not
```
Yes, I see the problem, I will think more deeper about such case.

clecat commented:
Those last commits are quite messy, they will know many changes

dinosaure commented on lib/de.ml:
```diff
@@ -1445,6 +1448,284 @@ module Inf = struct
+               (lit.Lookup.t.(d.hold land lit.Lookup.m) land Lookup.mask == 256
+               && lit.Lookup.t.(d.hold land lit.Lookup.m) lsr 15 <= d.bits
+               && i_rem d < 1)
+        then _fill_bits d in
```
May be inline `fill_bits` too, not sure that it's the right solution however.

clecat replied:
After checking with landmarks, it does not seem to be any better

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      ; flush_output os
+
+    let precode_lens_permutation =
+      [|16; 17; 18; 0; 8; 7; 9; 6; 10; 5; 11; 4; 12; 3; 13; 2; 14; 1; 15|]
```
Can be replaced by `zigzag` (see on top of `de.ml`).

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+       ; 26; 26; 26; 26; 26; 26; 26; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27
+       ; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27
+       ; 27; 27; 28
+      |]
```
Can be replaced by `_length` minor extra prepended `0`. Don't be afraid to update `_lenght` and the stream API to use the same array.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      [|
+         3; 4; 5; 6; 7; 8; 9; 10; 11; 13; 15; 17; 19; 23; 27; 31; 35; 43; 51; 59
+       ; 67; 83; 99; 115; 131; 163; 195; 227; 258
+      |]
```
Seems equal to `Array.map ((+) 3) _base_length`, can you re-use the same array?

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      [|
+         0; 0; 0; 0; 0; 0; 0; 0; 1; 1; 1; 1; 2; 2; 2; 2; 3; 3; 3; 3; 4; 4; 4; 4
+       ; 5; 5; 5; 5; 0
+      |]
```
See `_extra_lbits`.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      [|
+         1; 2; 3; 4; 5; 7; 9; 13; 17; 25; 33; 49; 65; 97; 129; 193; 257; 385; 513
+       ; 769; 1025; 1537; 2049; 3073; 4097; 6145; 8193; 12289; 16385; 24577
+      |]
```
Seems equal to `Array.map succ _base_dist`.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+      [|
+         0; 0; 0; 0; 1; 1; 2; 2; 3; 3; 4; 4; 5; 5; 6; 6; 7; 7; 8; 8; 9; 9; 10; 10
+       ; 11; 11; 12; 12; 13; 13
+      |]
```
See `_extra_dbits`.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+            ; incr i
+          done
+          ; f i (len - 1) in
+      f (ref 0) max_codeword_len
```
Hmmhmm, maybe translate the `i` argument to a simple integer instead of a `ref int` or translate the recursion to a loop with `ref`.

clecat replied:
Removing the recursion for a loop was the best option, so this is what I did

dinosaure replied:
But can you remove the ref?

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+              ; incr run_start
+            done
+          ; if !run_start <> num_lens then f () else !itemptr in
+        f ()
```
Not sure that the best should be to use a recursive function when you use reference inside, again two possibilities exist:
- re-implement it in a full functional way (without any reference)
- use a loop with reference

The mix of both (recursive function & reference) is hard to optimize then.

clecat replied:
Indeed, those are relics of when I was unsure about which one to choose, for now, I'll stick to loops

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+              done)
+        ; c.num_precode_items <-
+            compute_precode_items
+              (Array.append c.codes.lens.litlen c.codes.lens.offset)
```
It should be a bottle-neck but a dragoon appears here where `Array.append` has a real cost (allocation & copy).

clecat replied:
I had forgotten this one:
In the C codebase, the two arrays are next to each other inside a `struct`, therefore, the casting has no cost and allows to easily go trough both of them at the same time.
I had chosen to simply copy them in the first time, but it will need a thorough cleaning of the entire functions

dinosaure replied:
Ok, let's keep as is this part when it does not concern the hot-loop (compression / encoding) and focus on the hot-loop to really optimize the deflation :+1:.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+        done
+
+    let max_extra_length_bits = 5
+    let max_extra_offset_bits = 14
```
To homogenize with the rest of the code, constant should be prepended by a `_` :+1:.

clecat replied:
This is done, however, I'll have to come back to them, as we have several doublons with the streamable api

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+    let finish_sequence seq litrunlen = seq.litrunlen_and_length <- litrunlen
+    let num_observations_per_block_check = 512
+
+    let do_end_block_check stats block_length =
```
What is the purpose of this function? It tries to check if the input has a block tagged with last?

clecat replied:
This function contains some heuristic for choosing if a new block should be created or not

dinosaure commented on bench/dune:
```diff
@@ -57,3 +57,22 @@
+(library
+ (name de_landmarks)
+ (optional)
+ (modules de_landmarks)
```
You can add `(flags (:standard -w -55))` to keep some function with the `inline` tag (on the `de.ml`) and be able to compile the landmarks target without trouble.

clecat replied:
Thanks a lot, it's pushed

dinosaure commented:
![perf](https://snipboard.io/RpiTxY.jpg)

A simple perf on `bench/densld.exe` with `book1`: [*.svg](https://paste.isomorphis.me/lTV)

About the inflation, it seems that `fill_bits` it's the biggest bottle-neck. A possible solution is about the usage of our `hold`. Indeed, instead of execute the systematic mechanic where we set/save bits into your hold, we are in the situation where:
- we have the full input
- we know where we are in the input
- we still continue to update and use `d.bits` as our local counter

So instead of fill `hold` every time in our hot-loop, the best is to directly extract needed bits from the input according `i_pos` and `bits` for example.

An other possibility is to pass `hold` and `bits` on the stack and avoid the barrier limit between local functions in `inflate` - so `length`/`extra_length`/`distance`/`extra_distance`/`write` take 2 arguments & we should never call `caml_modify` in this area.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2892,1001 @@ module Def = struct
+        mf.hash4_tab
+      ; Array.iteri
+          (fun i v -> mf.next_tab.(i) <- v - (window_size land 0xFFFF))
+          mf.next_tab
```
may be instead of the use of `Array.iteri`, the best is to use a simple `for` loop.

dinosaure commented on lib/de.ml:
```diff
@@ -2605,6 +2892,1001 @@ module Def = struct
+            os.i_pos - !best_matchptr
+          else begin
+            _matchfinder_longest_rec cur_node best_matchptr os lens mf
+              depth_remaining cutoff
```
It seems that `best_matchptr` and `depth_remaining` are references on the heap. According to the `perf` graph:
![perf](https://snipboard.io/v5Hiy7.jpg)

We have a call to the GC and from the assembly generated, a call to `caml_call_gc` is about these values. Can you try to dereference them? At least you can dereference them and re-reference them then in `_matchfind_longest_rec` or try a fully functional implementation.

clecat commented:
About those last commits and @dinosaure  comments:
- For the inflation, I have solved the issues with fill_bits and removed the ring buffer: It was a relicate of the streamable API and was of no use there. It allowed me to be now strictly faster than the streamable API (with around 25 to 40 %).
- For the deflation, I changed the array of sequences into a list: It allows to avoid instanciating 10k elements sets to 0. It's benefits are instantly visible: We get sometime 600% of efficiency on small inputs, with only a negligeable drawback for some of the huge corpus.
- It could be a better thing to have a double linked list instead of a list, so we could avoid the reverse of the sequences

dinosaure commented:
Something is bad about 32-bits architecture, may be it's about `optint`. Then, can you rebase the PR with the current state of `decompress`? We have some conflicts.

clecat commented:
I've seen that, I'll resolve the conflicts

dinosaure commented:
Ok, so this PR remains 2 goals:
- [ ] clean the history
- [ ] show a final benchmark between `decompress`,`decompress (non-stream)` and `zlib`
- [ ] bug about remaining bytes

The support of `gzip` will be delayed - we need to test such code-path more deeply (with `ocaml-git`/`irmin`) before to propose something. However, it's a great work. We did not get too much improvements about performances but we did something better than the stream implementation. Thanks @clecat, I will do a pass after that to clean the repository according my view.

**NOTE:** we tested this PR with `ocaml-git` (see mirage/ocaml-git#502) & irmin. As far as we can tell, we did not notice any bugs.

dinosaure commented:
The current state about performance:
```sh
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  19437 MB/s 20362 MB/s      111261 100.00 /home/dinosaure/dev/decompress/test/corpus/bib
decompress_ns dev -4       20 MB/s   108 MB/s       36251  32.58 /home/dinosaure/dev/decompress/test/corpus/bib
decompress dev -4          16 MB/s    84 MB/s       44072  39.61 /home/dinosaure/dev/decompress/test/corpus/bib
zlib 1.2.11 -4             39 MB/s   224 MB/s       37555  33.75 /home/dinosaure/dev/decompress/test/corpus/bib
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  14079 MB/s 14843 MB/s      768771 100.00 /home/dinosaure/dev/decompress/test/corpus/book1
decompress_ns dev -4       14 MB/s    81 MB/s      319056  41.50 /home/dinosaure/dev/decompress/test/corpus/book1
decompress dev -4          12 MB/s    60 MB/s      379395  49.35 /home/dinosaure/dev/decompress/test/corpus/book1
zlib 1.2.11 -4             30 MB/s   205 MB/s      329150  42.82 /home/dinosaure/dev/decompress/test/corpus/book1
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  14604 MB/s 14857 MB/s      610856 100.00 /home/dinosaure/dev/decompress/test/corpus/book2
decompress_ns dev -4       17 MB/s   105 MB/s      211710  34.66 /home/dinosaure/dev/decompress/test/corpus/book2
decompress dev -4          15 MB/s    76 MB/s      259394  42.46 /home/dinosaure/dev/decompress/test/corpus/book2
zlib 1.2.11 -4             35 MB/s   220 MB/s      217629  35.63 /home/dinosaure/dev/decompress/test/corpus/book2
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  14375 MB/s 14953 MB/s      377109 100.00 /home/dinosaure/dev/decompress/test/corpus/news
decompress_ns dev -4       18 MB/s    95 MB/s      147129  39.01 /home/dinosaure/dev/decompress/test/corpus/news
decompress dev -4          13 MB/s    71 MB/s      174833  46.36 /home/dinosaure/dev/decompress/test/corpus/news
zlib 1.2.11 -4             32 MB/s   199 MB/s      149282  39.59 /home/dinosaure/dev/decompress/test/corpus/news
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  14787 MB/s 14854 MB/s      513216 100.00 /home/dinosaure/dev/decompress/test/corpus/pic
decompress_ns dev -4       36 MB/s   215 MB/s       57160  11.14 /home/dinosaure/dev/decompress/test/corpus/pic
decompress dev -4          31 MB/s   206 MB/s       70467  13.73 /home/dinosaure/dev/decompress/test/corpus/pic
zlib 1.2.11 -4             73 MB/s   377 MB/s       58457  11.39 /home/dinosaure/dev/decompress/test/corpus/pic
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
```

What that means?
- `decompress` (stream) is 3 times slower than `zlib` about compression (nothing change)
- `decompress` (non-stream) is 2 times slower than `zlib` which is good :+1:
- `decompress` (non-stream) has a better ratio than `zlib` or `decompress` (stream) - which is really good
- about decompression, factors are globally the same & `decompress`(non-stream) is better than `decompress` (stream)

These tests are done on the Calgary corpus. I think it a really good result and we should introduce it at the next release :+1:. Bravo @clecat! It's not an easy work. We will iterate then about performances later but it's nice first shot.

dinosaure commented:
Force push and rebase on the current state of `decompress~master`.

dinosaure commented:
So I think it's time to merge it and prepare a release of `decompress.1.4.0`. Thanks for this great work, at least, we can observe an improvement about performances and ratio which is good and expected.
